eng/versioning/update_versions.py

```diff
@@ -0,0 +1,194 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+# Use case: Update all the versions in README.md and pom.xml files based on
+# the versions in versions_[client|data|management].txt/dependencies_[client|data|management].txt
+#
+#    python eng/versioning/update_versions.py --update-type [library|external_dependency|all] -build-type [client|data|management]
+# For example: To update the library versions for the client track
+#    python eng/versioning/update_versions.py --ut library --bt client
+#
+# Use case: Update the versions in a particular file
+#
+#    python utilities/update_versions.py --update-type [library|external_dependency|all] -build-type [client|data|management] --target-file pom-file-to-update
+# For example: To update all versions for the client track for a given pom file
+#    python eng/versioning/update_versions.py --ut all --bt client --tf <pathToPomFile>\pom.xml
+#
+# The script must be run at the root of azure-sdk-for-java.
+
+import argparse
+from datetime import timedelta
+from enum import Enum
+import os
+import re
+import time
+
+version_update_start_marker = re.compile(r'\{x-version-update-start;([^;]+);([^}]+)\}')	
+version_update_end_marker = re.compile(r'\{x-version-update-end\}')
+version_update_marker = re.compile(r'\{x-version-update;([^;]+);([^}]+)\}')
+# regex for the version string is suggested one directly from semver.org
+# it's worth noting that both regular expressions on that page have start
+# of line (^) and end of line ($) anchors which need to be removed since
+# what's being matched is in the middle of the string
+# https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
+version_regex_str = r'(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?'
+
+class UpdateType(Enum):
+    external_dependency = 'external_dependency'
+    library = 'library'
+    all = 'all'
+
+    # defining string is necessary to get ArgumentParser's help output to produce
+    # human readable values of UpdateType
+    def __str__(self):
+        return self.value
+
+class BuildType(Enum):
+    client = 'client'
+    data = 'data'
+    management = 'management'
+    
+    # defining string is necessary to get ArgumentParser's help output to produce
+    # human readable values of BuildType
+    def __str__(self):
+        return self.value
+
+class CodeModule:
```

Comment:
Looks like a fair bit of this logic is duplicated across scripts so it would be good to share that in a future clean-up.

Reply:
The 12th commit moves some common classes and the regex strings into a utils file. I also had to add .gitignore since importing from a common file ends up creating a python cache "__pycache__" directory.